### PR TITLE
planner: reuse the range of indexes

### PR DESCRIPTION
**Pull request by winoros**
### What problem does this PR solve? 

reuse the range of indexes to reduce the CPU cost.

### What is changed and how it works?

Pass the `accessPath` to `Selectivity` so that we don't need build new ranges if there's a path to cover it.

### Check List 

Tests 

 - Existing Unit test

Code changes

 - Has exported function/method change
 - Has exported variable/fields change

Side effects

 - Increased code complexity

Related changes

 - Need to cherry-pick to the release branch (should be decided later)

### CODE REVIEW

File: cmd/explaintest/r/explain_easy.result [hunk 1]
```diff
@@ -76,9 +76,9 @@ TopN_7	1.00	root	test.t2.c2:asc, offset:0, count:1
     └─TableScan_13	10000.00	cop	table:t2, range:[-inf,+inf], keep order:false, stats:pseudo
 explain select * from t1 where c1 > 1 and c2 = 1 and c3 < 1;
 id	count	task	operator info
-IndexLookUp_11	1.11	root	
+IndexLookUp_11	11.08	root
```
Comment (winoros) [inline cmd/explaintest/r/explain_easy.result]: https://github.com/pingcap/tidb/pull/10984/files#diff-20ac3c94a9605bc56ae644dfd45569f8R234 changes the row count estimation. Maybe we need to first support the pk in index's histogram.

File: planner/core/cbo_test.go [hunk 2]
```diff
@@ -1071,10 +1071,9 @@ func (s *testAnalyzeSuite) TestLimitCrossEstimation(c *C) {
 	tk.MustExec("insert into t values (1, 1),(2, 1),(3, 1),(4, 1),(5, 1),(6, 1),(7, 1),(8, 1),(9, 1),(10, 1),(11, 1),(12, 1),(13, 1),(14, 1),(15, 1),(16, 1),(17, 1),(18, 1),(19, 1),(20, 2),(21, 2),(22, 2),(23, 2),(24, 2),(25, 2)")
 	tk.MustExec("analyze table t")
 	tk.MustQuery("EXPLAIN SELECT * FROM t WHERE b = 2 and a > 0 ORDER BY a limit 1").Check(testkit.Rows(
-		"TopN_8 1.00 root test.t.a:asc, offset:0, count:1",
-		"└─IndexReader_16 1.00 root index:TopN_15",
-		"  └─TopN_15 1.00 cop test.t.a:asc, offset:0, count:1",
-		"    └─IndexScan_14 7.50 cop table:t, index:b, range:(2 0,2 +inf], keep order:false",
+		"TopN_10 1.00 root test.t.a:asc, offset:0, count:1",
```
Comment (winoros) [inline planner/core/cbo_test.go]: Same reason.

**Comment by winoros** [comment]
WechatIMG9

Before this commit, you can see that there's a scenario that the range cost is very very high. By reuse the range, we can reduce the cost of `DetachCondAndBuildRangeForIndex` in `getMaskAndRanges`.

**Comment by zz-jason** [comment]
@winoros Please update the master branch and resolve conflicts.

**Comment by winoros** [comment]
After we support the `multiple index read`. The scenarios become more complicated.
This PR's content needs to be refined.
Closed for now.

**Pull request closed without merging**